### PR TITLE
Perform protocol versioning and merge packages

### PR DESCRIPTION
**Pull request by msujew**
I've recently noticed that there's quite a bit of information leakage from the protocol to the underlying messaging infrastructure. Mainly IDs of peers. Given that the RPC and protocol are so intertwined, it makes sense to put them into the same package and give them a single versioning number.

Each client will now check whether the used client version satisfies the protocol version required by the server.

### CODE REVIEW

File: packages/open-collaboration-protocol/src/connection-provider.ts [hunk 1]
```diff
@@ -76,12 +82,28 @@ export class ConnectionProvider {
...
+        const metadata = await this.getMetaData();
+        const serverVersion = semver.parse(metadata.version);
+        if (!serverVersion) {
+            throw new Error('Invalid protocol version returned by server: ' + metadata.version);
+        }
+        if (!semver.valid(this.protocolVersion)) {
+            throw new Error('Invalid protocol version provided by client: ' + this.protocolVersion);
+        }
+        // Check if the server version is compatible with the client version
+        const serverVersionRange = `^${serverVersion.format()}`;
```
Comment (spoenemann) [inline packages/open-collaboration-protocol/src/connection-provider.ts:95]: We could analyze the major, minor and patch components of `serverVersion` to construct a version that has only the first non-zero component, e.g. `0.1.2` > `^0.1.0`, `2.1.5` > `^2.0.0`

That way we could eliminate the need for the additional constant `MIN_VERSION` returned by the server.

File: packages/open-collaboration-server/src/collaboration-server.ts [hunk 2]
```diff
@@ -203,8 +206,8 @@ export class CollaborationServer {
         });
         app.get('/api/meta', async (_, res) => {
             const data: types.ProtocolServerMetaData = {
-                owner: '',
-                version: '',
+                owner: process.env['OCT_SERVER_OWNER'] ?? 'Unknown',
```
Comment (spoenemann) [inline packages/open-collaboration-server/src/collaboration-server.ts:209]: How this this meant to be used? Should it contain a company name?

Please use the configuration service to access the env variable.

File: packages/open-collaboration-vscode/src/commands.ts [hunk 3]
```diff
@@ -69,16 +69,20 @@ export class Commands {
...
-                    } else if (index === 1) {
-                        await this.roomService.joinRoom(connectionProvider);
+                    try {
+                        if (index === 0) {
+                            await this.roomService.createRoom(connectionProvider);
+                        } else if (index === 1) {
+                            await this.roomService.joinRoom(connectionProvider);
+                        }
+                    } catch (error) {
+                        vscode.window.showErrorMessage('Failed to create or join room: ' + String(error));
```
Comment (spoenemann) [inline packages/open-collaboration-vscode/src/commands.ts:79]: The message should say either 'create' or 'join' depending on the user's selection.

**Review by spoenemann — COMMENTED** [review]
Thanks, the change looks good overall.

Q: Should we also check the `version` field of received messages, in the server or client or both?

File: packages/open-collaboration-server/src/collaboration-server.ts (hunk 2 above)
Comment (msujew) [reply packages/open-collaboration-server/src/collaboration-server.ts]: I'd put the company name in there, yes 👍

File: packages/open-collaboration-protocol/src/connection-provider.ts (hunk 1 above)
Comment (msujew) [reply packages/open-collaboration-protocol/src/connection-provider.ts]: I've added a method that can be used by both the server and the client to reject invalid versions.

File: packages/open-collaboration-protocol/src/connection-provider.ts [hunk 4]
```diff
@@ -33,11 +38,21 @@ export class ConnectionProvider {
...
         this.fetch = options.fetch ?? ((url, options) => fetch(url, options));
         this.userAuthToken = options.userToken;
+        if (options.protocolVersion) {
+            const parsed = semver.parse(options.protocolVersion);
+            if (!parsed) {
+                throw new Error('Invalid protocol version provided: ' + options.protocolVersion);
+            }
+            this.protocolVersion = parsed;
+        } else {
+            this.protocolVersion = new semver.SemVer(VERSION);
```
Comment (spoenemann) [inline packages/open-collaboration-protocol/src/connection-provider.ts:54]: You can use `SEM_VERSION` here.

File: packages/open-collaboration-server/src/collaboration-server.ts [hunk 5]
```diff
@@ -203,8 +206,8 @@ export class CollaborationServer {
         });
         app.get('/api/meta', async (_, res) => {
             const data: types.ProtocolServerMetaData = {
-                owner: '',
-                version: '',
+                owner: this.configuration.getValue('oct-server-owner') ?? 'Unknown',
```
Comment (spoenemann) [inline packages/open-collaboration-server/src/collaboration-server.ts:209]: Please add this config variable to the server README.

**Review by spoenemann — APPROVED** [review]
🚀

**Review by spoenemann — APPROVED** [review]

**Review by spoenemann — COMMENTED** [review]
Oh, now I get "Activating extension 'typefox.open-collaboration-tools' failed: self is not defined." in the vscode extension

**Review by spoenemann — APPROVED** [review]
Working again, thanks!

File: packages/open-collaboration-protocol/src/connection-provider.ts [hunk 6]
```diff
@@ -129,6 +129,9 @@ export class ConnectionProvider {
                 'x-oct-jwt': this.userAuthToken!
             }
         });
+        if (response.status !== 200) {
```
Comment (spoenemann) [inline packages/open-collaboration-protocol/src/connection-provider.ts:132]: There's a `response.ok` flag for this (all codes from 200 to 299 are considered ok).

Comment (msujew) [reply packages/open-collaboration-protocol/src/connection-provider.ts:132]: Thanks, see https://github.com/TypeFox/open-collaboration-tools/pull/31/commits/880474003210bdf856c09255e9f0658215654fb1.

**Pull request merged**